### DISCUSSION TITLE
merge_with and arena usage

### DISCUSSION THREAD

**Discussion by KimBP** [biojppm/rapidyaml#306] [Q&A] [answered]
I'm trying to use rapidYaml to parse a settings file required by my application.
Basically the structure is like this:
```
ryml::Tree tree; // Global tree containing my settings

set_factory_default();
get_from_settings_file():
```
In set_factory_default I build my full settings tree with default values.
In get_from_settings_file() I open the settings file and parse it into a local (stack variable) tree.
```
ryml::Tree filetree = ryml::parse_in_place(ryml::to_substr(filedata));
tree.merge_with(&filetree);
```

When I return from get_from_settings_file() tree is updated with 'overrides' from my filedata. Unfortunately the first key-val pair read from filedata is corrupted. I think it is due to the substr and/or filetree object being destroyed when returning (i.e. stack is reused for other stuff)
I've found one solution (which I do consider a none solution)

Make filetree global too and use
filetree = ryml::parse_in_arena(ryml::to_csubstr(filedata);

Isn't there a way to ensure data follows the tree I'm merging into?

**Comment by biojppm** [comment, marked as answer]
(Sorry for the late reply, Github's notifications are just broken for me).

Indeed, any data added to `tree` needs to go on existing for as long as `tree` does. In this case you are adding from `filedata`, so you need to make sure that it exists with `tree`. There are several ways, but one simple way is to do `filedata = tree.alloc_arena(filesize)` on the destination tree, then still do the parse in place into the temporary `filetree`.
```c++
ryml::Tree tree;
set_factory_default();

{
    size_t filesize = ...;
    ryml::substr filedata = tree.alloc_arena(filesize); // place the filedata in the destination tree
    file_get_contents(file, filedata);
    ryml::Tree filetree = ryml::parse_in_place(filedata); // read the temp tree in place
    tree.merge_with(&filetree);
}

// at this point, filetree's contents are safely in the final tree.
```
For a similar example, see [this note](https://github.com/biojppm/rapidyaml/issues/288#issuecomment-1200579112).

HTH.

**Comment by biojppm** [comment]
Also, you may want to take a look at [c4conf](https://github.com/biojppm/c4conf), which does exactly what you are looking for.
